File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/internal/NodeOutputImpl.java
package io.littlehorse.sdk.wfsdk.internal;

import io.littlehorse.sdk.common.exception.LHMisconfigurationException;
import io.littlehorse.sdk.common.proto.LHPath.Selector;
import io.littlehorse.sdk.wfsdk.NodeOutput;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import lombok.Setter;

class NodeOutputImpl implements NodeOutput {

    public String nodeName;
    public WorkflowThreadImpl parent;

    @Getter
    @Setter
    private String jsonPath;

    @Getter
    private List<Selector> lhPath;

    public NodeOutputImpl(String nodeName, WorkflowThreadImpl parent) {
        this.nodeName = nodeName;
        this.parent = parent;
        this.lhPath = new ArrayList<>();
    }

    public NodeOutputImpl jsonPath(String path) {
        if (jsonPath != null) {
            throw new RuntimeException("Cannot use jsonpath() twice on same node!");
        }
        NodeOutputImpl out = new NodeOutputImpl(nodeName, parent);
        out.setJsonPath(path);
        return out;
    }

    public NodeOutputImpl get(String index) {
        if (jsonPath != null) {
            throw new LHMisconfigurationException("Cannot use jsonPath() and get() on same var!");
        }
        NodeOutputImpl out = new NodeOutputImpl(nodeName, parent);
        out.getLhPath().add(Selector.newBuilder().setKey(index).build());
        return out;
    }
}

File: sdk-js/src/workflow/internal/nodeOutputImpl.ts
import type { VariableAssignment } from '../../proto/common_wfspec'
import { VariableMutationType } from '../../proto/common_wfspec'
import { LHMisconfigurationException } from '../exceptions'
import type { NodeOutput } from '../nodeOutput'
import type { WorkflowRhs } from '../workflowRhs'
import { LHExpressionImpl } from './lhExpressionImpl'
import type { WorkflowThreadImpl } from './workflowThreadImpl'

export class NodeOutputImpl implements NodeOutput {
  readonly jsonPathValue?: string
  readonly lhSelectors: { selectorType?: { $case: 'key'; value: string } | { $case: 'index'; value: number } }[]

  constructor(
    readonly nodeName: string,
    readonly thread: WorkflowThreadImpl,
    jsonPath?: string,
    lhSelectors?: NodeOutputImpl['lhSelectors']
  ) {
    this.jsonPathValue = jsonPath
    this.lhSelectors = lhSelectors ?? []
  }

  toAssignment(): VariableAssignment {
    const path =
      this.jsonPathValue !== undefined
        ? { $case: 'jsonPath' as const, value: this.jsonPathValue }
        : this.lhSelectors.length > 0
          ? { $case: 'lhPath' as const, value: { path: [...this.lhSelectors] } }
          : undefined
    return {
      path,
      source: {
        $case: 'nodeOutput',
        value: { nodeName: this.nodeName },
      },
      targetType: undefined,
    }
  }

  jsonPath(path: string): NodeOutput {
    if (this.jsonPathValue !== undefined) {
      throw new LHMisconfigurationException('Cannot use jsonPath() twice on the same node output')
    }
    if (this.lhSelectors.length > 0) {
      throw new LHMisconfigurationException('Cannot use jsonPath() and get() on the same node output')
    }
    return new NodeOutputImpl(this.nodeName, this.thread, path, [])
  }

  get(field: string): NodeOutput {
    if (this.jsonPathValue !== undefined) {
      throw new LHMisconfigurationException('Cannot use jsonPath() and get() on the same node output')
    }
    const next = [...this.lhSelectors, { selectorType: { $case: 'key' as const, value: field } }]
    return new NodeOutputImpl(this.nodeName, this.thread, undefined, next)
  }

  add(other: WorkflowRhs): LHExpressionImpl {
    return new LHExpressionImpl(this, VariableMutationType.ADD, other)
  }

  subtract(other: WorkflowRhs): LHExpressionImpl {
    return new LHExpressionImpl(this, VariableMutationType.SUBTRACT, other)
  }

  multiply(other: WorkflowRhs): LHExpressionImpl {
    return new LHExpressionImpl(this, VariableMutationType.MULTIPLY, other)
  }

  divide(other: WorkflowRhs): LHExpressionImpl {
    return new LHExpressionImpl(this, VariableMutationType.DIVIDE, other)
  }
}
